marimbabot_vision/src/vision_node.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python
+
+import rospy
+import torch
+from cv_bridge import CvBridge, CvBridgeError
+from PIL import Image as PILImage
+from sensor_msgs.msg import Image as ROSImage
+from transformers import (DonutProcessor, VisionEncoderDecoderConfig,
+                          VisionEncoderDecoderModel)
+
+
+def detect_notes(open_cv_img, pre_processor: DonutProcessor, model: VisionEncoderDecoderModel):
+    # Load image
+    image = PILImage.fromarray(open_cv_img).convert('RGB')
+
+    # Rotate image if needed
+    if image.size[0] > image.size[1]:
+        image = image.transpose(PILImage.Transpose.ROTATE_90)
+
+    # Generate initial sequence
+    decoder_input_ids = pre_processor.tokenizer(
+        "<s>",
+        add_special_tokens=False,
+        return_tensors="pt").input_ids
+
+    # Preprocess image
+    pixel_values = pre_processor(image, return_tensors="pt").pixel_values
+
+    # Run the model
+    outputs = model.generate(
+        pixel_values.to(device),
+        decoder_input_ids=decoder_input_ids.to(device),
+        max_length=model.decoder.config.max_position_embeddings,
+        early_stopping=True,
+        pad_token_id=pre_processor.tokenizer.pad_token_id,
+        eos_token_id=pre_processor.tokenizer.eos_token_id,
+        use_cache=True,
+        num_beams=1,
+        bad_words_ids=[[pre_processor.tokenizer.unk_token_id]],
+        return_dict_in_generate=True,
+    )
+
+    # Decode tokens
+    sequence = pre_processor.batch_decode(outputs.sequences, skip_special_tokens=True)[0]
+
+    print(sequence)
```

Comment:
These two lines print the output two times. I would remove them or at least lower the log level to debug. Otherwise the terminal gets spamed quite a bit and one can always easily subscribe to the output topic.